### PR TITLE
[draft] flow callbacks for ndpi - v1

### PR DESCRIPTION
**Pull request by jasonish**
In order for nDPI to become a plugin it needs to hook into flow initialization and flow updates. This PR provides user registerable flow callbacks for these events.

It also includes a "dummy" ndpi plugin using these callbacks, and also shows how to use the flow storage API to avoid modifying the Flow structure directly.

Original nDPI PR: #11671 

Tickets:
- https://redmine.openinfosecfoundation.org/issues/7319
- https://redmine.openinfosecfoundation.org/issues/7320

### CODE REVIEW

File: plugins/ndpi-dummy/ndpi.c [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+static void FlowStorageFree(void *ptr)
+{
+    SCLogNotice("De-allocating nDPI flow storage");
+    int *dummy_storage = ptr;
+    SCLogNotice("%d", *dummy_storage);
+    SCFree(ptr);
+}
+
+static void OnFlowInit(Flow *f, const Packet *p)
```
Comment (victorjulien) [inline plugins/ndpi-dummy/ndpi.c:45]: I guess I'd expect a OnFlowDestruct too, although perhaps we don't need it directly here, as the FlowStorage API knowns how to destroy the user data.

File: plugins/ndpi-dummy/ndpi.c [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+{
+    SCLogNotice("...");
+    static int counter = 0;
+    int *dummy_storage = SCCalloc(1, sizeof(int));
+    *dummy_storage = counter++;
+    FlowSetStorageById(f, flow_storage_id, dummy_storage);
+}
+
+static void OnFlowUpdate(Flow *f, Packet *p, ThreadVars *tv)
```
Comment (victorjulien) [inline plugins/ndpi-dummy/ndpi.c:54]: should there be user data here as well? general user data? Perhaps not needed for ndpi

Comment (victorjulien) [reply plugins/ndpi-dummy/ndpi.c:54]: I guess more for the library, but if someone wanted to create a simple flow logger, they'd want to pass a file pointer around these callbacks to log a new flow creation, or a flow destruction, etc.

File: plugins/ndpi-dummy/ndpi.c (hunk 1 above)
Comment (jasonish) [reply plugins/ndpi-dummy/ndpi.c:45]: Yeah, for completeness should probably add it. Not needed for storage cleanup using the storage API, and someone else uses the flow event for their private cleanup, but it does feel missing.

File: plugins/ndpi-dummy/ndpi.c (hunk 2 above)
Comment (jasonish) [reply plugins/ndpi-dummy/ndpi.c:54]: Probably should have user data, was being short sighted and only looking at the case at hand which doesn't need it.

**Comment by jasonish** [comment]
Replaced by https://github.com/OISF/suricata/pull/11942, all callbacks accept user data now.

**Pull request closed without merging**